beancount_import/matching.py

```diff
@@ -255,6 +283,81 @@ def add_posting(self, entry: Transaction, mp: MatchablePosting):
         group = self._postings.setdefault(_date_key(entry, mp), {})
         group[source_posting_ids] = (entry, mp)
 
+        weight = get_posting_weight(mp.posting)
+        if weight is None:
+            return
+
+        for dc in self.fuzz_date_currency_key(self._date_currency_key(entry, mp)):
+            bisect.insort(
+                self._date_currency[dc],
+                self._search_posting(entry, mp))
+
+    def search_postings(self,
+                        entry: Transaction,
+                        mps: List[MatchablePosting],
+                        cmp_callable: Callable,
+                        negate=False) -> Iterable[SearchPosting]:
+        # Prepare postings for searching
+        postings_date_currency = collections.defaultdict(list)
+        
+        for mp in mps:
+            weight = get_posting_weight(mp.posting)
+            if weight is None:
+                continue
+            if negate:
+                weight = -weight
+            bisect.insort(
+                postings_date_currency[self._date_currency_key(entry, mp)],
+                (weight, id(mp), mp.posting))
+
+        keys = sorted(postings_date_currency.keys())
+
+        for key in keys:
+            db = self._date_currency[key]
+            s = slice(0,1)
+            for weight, _, p in postings_date_currency[key]:
+                # Do meta lookup first
+                if not negate and not is_unknown_account(p.account):
+                    meta = p.meta
+                    if meta:
+                        # Look for metadata matches
+                        yielded = False
+                        for meta_key in self.metadata_keys:
+                            value = meta.get(meta_key)
+                            if value is None: continue
+                            cur_matches = self._get_weight_matches(
+                                weight, self._keyed_postings.get(
+                                    (posting.account, meta_key, value)))
+                            if cur_matches:
+                                for (k, (t, mp)) in cur_matches.items():
+                                    # Stop searching if we match a keyed posting
+                                    newsp = SearchPosting(
+                                        number=None, key=k, entry=t, mp=mp)
+                                    if cmp_callable(p, newsp):
+                                        yielded = True
+                                        yield newsp
+                        if yielded:
+                            continue
+
+                upper = weight.number + self.fuzzy_match_amount
```

Comment:
I don't think there is necessarily much gain in trying to do multiple posting queries as a batch here.  I think if you just did a binary search it would be plenty fast and would simplify the code --- basically the same structure as before could be used.

Reply:
I ran both implementations side by side and was surprised to find that you're right :)

I'm working on removing the new code I created and adopting the bisect implementation.

Reply:
After fiddling with the code some more and clearing some bugs, it turns out I was wrong.

My bug was I was always finding 0 results during matching. Turns out that's faster than actually matching things.

Current implementation wise, the bisecting takes 2.1 seconds (time spent in _get_matches), the sliding iterators take 0.93 seconds (time spent in search_postings).